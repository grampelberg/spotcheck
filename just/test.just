[no-cd]
run *ARGS:
    # Force `./src` because bun test doesn't respect package.json
    bun test {{ ARGS }} src

[no-cd]
dev $LOG_LEVEL="*=debug":
    just test::run "--watch"

[no-cd]
concurrent:
    just test::run "--concurrent"

default: run
