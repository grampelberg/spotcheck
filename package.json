{
  "name": "@grampelberg/spotcheck",
  "type": "module",
  "scripts": {
    "dep:check": "depcheck",
    "format:check": "biome format && just --fmt --unstable --check",
    "lint:check": "biome lint --error-on-warnings && tsc --noEmit"
  },
  "exports": {
    ".": "./src/lib.ts",
    "./bun": {
      "default": "./src/bun.ts",
      "types": "./types/bun.d.ts"
    }
  },
  "devDependencies": {
    "depcheck": "^1.4.7"
  },
  "peerDependencies": {
    "@types/bun": "latest",
    "typescript": "^5"
  },
  "dependencies": {
    "@types/he": "^1.2.3",
    "@types/pngjs": "^6.0.5",
    "@types/debug": "^4.1.12",
    "callsites": "^4.2.0",
    "chalk": "^5.6.2",
    "debug": "^4.4.3",
    "generic-pool": "^3.9.0",
    "he": "^1.2.0",
    "pixelmatch": "^7.1.0",
    "pngjs": "^7.0.0",
    "puppeteer": "^24.22.3",
    "react": "^19.1.1",
    "react-dom": "^19.1.1",
    "sharp": "^0.34.4",
    "zod": "^4.1.11"
  },
  "peerDependenciesMeta": {
    "@types/bun": {
      "optional": true
    }
  }
}
