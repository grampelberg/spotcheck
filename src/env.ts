import { z } from 'zod'

// biome-ignore lint/nursery/useExplicitType: zod
const envBool = z.preprocess(val => {
  if (val === 'true' || val === 1 || val === 'yes') return true
  if (val === '' || val === 'false' || val === 0 || val === 'no') return false
  throw new Error(`Invalid boolean value: ${val}`)
}, z.boolean())

// biome-ignore lint/nursery/useExplicitType: zod
const settingsSchema = z.object({
  SPOTCHECK_PRESERVE: envBool.default(false),
  SPOTCHECK_UPDATE: envBool.default(false),
  TEST_CLEAN: envBool.default(true),
})

export type Env = z.infer<typeof settingsSchema>

export const env = (): Env => settingsSchema.parse(process.env)
