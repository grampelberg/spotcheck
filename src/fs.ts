import type { EncodingOption } from 'node:fs'
import fs from 'node:fs/promises'
import path, { join } from 'node:path'

import callsites from 'callsites'
import debug from 'debug'

import { name as pkgName } from '../package.json' with { type: 'json' }

const _log: debug.Debugger = debug(`${pkgName}:lib`)

interface MkdtempResult {
  path: string
  remove: () => Promise<void>
  [Symbol.asyncDispose]: () => Promise<void>
}

export const mkdtemp = async (
  prefix: string,
  options?: EncodingOption,
): Promise<MkdtempResult> => {
  const path = await fs.mkdtemp(prefix, options)

  const remove = async (): Promise<void> => {
    await fs.rmdir(path, {
      maxRetries: 0,
      recursive: true,
      retryDelay: 0,
    })
  }

  return {
    path,
    remove,
    async [Symbol.asyncDispose](): Promise<void> {
      await remove()
    },
  }
}

const callerDir = (): string => {
  // TODO: util.getCallSites() would be the best way to go about this.
  // Unfortunately, bun has not implemented it yet. For now, this assumes that
  // it is being run from within bun:test.
  const stack = callsites()

  const top = stack
    .slice()
    .reverse()
    .find(s => s.getFileName())
  if (!top) throw new Error('Could not determine caller')

  return path.dirname(top.getFileName() as string)
}

export const dirname = (loc: string): string => join(callerDir(), loc)

interface BaseOpts {
  idx?: number
  variant?: string
}

export const basename = (name: string, { idx, variant }: BaseOpts): string => {
  let fname = `${encodeURIComponent(name)}.${process.platform}`
  if (idx !== undefined) fname += `.${idx}`
  if (variant) fname += `.${variant}`
  fname += '.png'
  return fname
}
