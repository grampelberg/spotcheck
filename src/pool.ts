import type { Pool } from 'generic-pool'
import genericPool from 'generic-pool'
import puppeteer, { type Browser, type LaunchOptions } from 'puppeteer'
import { z } from 'zod'

import { env } from './env.ts'

const launchArgs: LaunchOptions = env().GITHUB_ACTIONS
  ? {
      args: ['--no-sandbox', '--disable-setuid-sandbox'],
    }
  : {}

// biome-ignore lint/nursery/useExplicitType: zod
const optSchema = z.object({
  max: z
    .number()
    .min(1)
    .default(10)
    .describe('Maximum number of browsers to spawn'),
  preserveBrowser: z
    .boolean()
    .default(false)
    .describe('Preserve browsers after use'),
})

export type PoolOptions = z.input<typeof optSchema>
type OOptions = z.output<typeof optSchema>

export const newPool = (rawOpts?: PoolOptions): genericPool.Pool<Browser> => {
  const opts: OOptions = optSchema.parse(rawOpts ?? {})
  opts.preserveBrowser = rawOpts?.preserveBrowser ?? env().SPOTCHECK_PRESERVE

  const factory = {
    create: (): Promise<Browser> =>
      puppeteer.launch({ headless: !opts.preserveBrowser, ...launchArgs }),
    destroy: (browser: Browser): Promise<void> => {
      if (!opts.preserveBrowser) return browser.close()
      return Promise.resolve()
    },
  }

  return genericPool.createPool(factory, { min: 1, max: opts.max })
}

interface DisposableBrowser {
  browser: Browser
  release: () => Promise<void>
  [Symbol.asyncDispose]: () => Promise<void>
}

export const acquire = async (
  pool: Pool<Browser>,
): Promise<DisposableBrowser> => {
  const browser = await pool.acquire()

  const release = async (): Promise<void> => await pool.release(browser)

  return {
    browser,
    release,
    async [Symbol.asyncDispose](): Promise<void> {
      release()
    },
  }
}
