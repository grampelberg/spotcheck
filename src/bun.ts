import type { CustomMatcher, MatcherResult } from 'bun:test'
import { afterAll, beforeAll } from 'bun:test'
import { mkdtemp, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import type { BunPlugin } from 'bun'
import chalk from 'chalk'
import debug from 'debug'
import { decode } from 'he'
import type { ReactElement } from 'react'

import { name as pkgName } from '../package.json' with { type: 'json' }
import type { BunOptions } from '../types/bun.ts'
import { env } from './env.ts'
import { basename, mkdtemp as mkdtempDisposable } from './fs.ts'
import type { Builder, IOptions, ScreenshotDiff } from './lib.ts'
import { optionsSchema, screenshot } from './lib.ts'
import { newPool } from './pool.ts'

const log: debug.Debugger = debug(`${pkgName}:bun`)

let diffDir: string = ''

beforeAll(async () => {
  diffDir = await mkdtemp(join(tmpdir(), `${pkgName}-`))
})

const builder =
  (plugins: BunPlugin[]): Builder =>
  async (index: string, css?: string[]): Promise<string[]> => {
    await using loc = await mkdtempDisposable(join(tmpdir(), 'screenshot-'))
    const indexPath = join(loc.path, 'index.html')
    await Bun.file(indexPath).write(decode(index))

    const build = await Bun.build({
      entrypoints: [indexPath, ...(css ?? [])],
      outdir: join(loc.path, 'dist'),
      naming: '[name]-[hash].[ext]',
      plugins,
    })

    if (!build.success) {
      throw new Error(`Build failed:\n${build.logs}`)
    }

    return await Promise.all(
      build.outputs
        .filter(f => f.path.endsWith('.css'))
        .map(async f => await Bun.file(f.path).text()),
    )
  }

const msgLine = (r: ScreenshotDiff, msg: string): string =>
  `${[r.state, r.idx].join('.').padEnd(10, ' ')}: ${msg}`

export const toMatchScreenshot = (
  globalOpts?: BunOptions,
): CustomMatcher<unknown, [string]> => {
  let pool = newPool(globalOpts?.pool)
  let isDrained = false

  afterAll(async () => {
    // TODO: this is a hack until [23113](https://github.com/oven-sh/bun/pull/23113)
    // is released.
    isDrained = true

    await pool.drain()
    await pool.clear()
  })

  return async (
    element: unknown,
    // Can go away once bun implements `expect.getState`
    name: string,
    rawOpts?: IOptions,
  ): Promise<MatcherResult> => {
    const callOpts = optionsSchema.parse(rawOpts ?? {})
    const opts = { ...globalOpts, ...callOpts }
    const plugins = opts.plugins ?? []

    opts.update = rawOpts?.update ?? env().SPOTCHECK_UPDATE

    log(`taking screenshot for "${name}":`, opts)

    // TODO: this is a hack, see above
    if (isDrained) {
      pool = newPool(globalOpts?.pool)
      isDrained = false
    }

    if (element === window.document) {
      element = (element as Document).body.outerHTML
    }

    const results = await screenshot(
      element as ReactElement | string,
      name,
      builder(plugins),
      pool,
      opts,
    )

    const matchResults = await Promise.all(
      results.map(async r => {
        if (!r.before) {
          return {
            message: msgLine(
              r,
              'No screenshot to compare against, generated one',
            ),
            pass: true,
          }
        }

        if (r.identical) {
          return {
            message: msgLine(r, 'Screenshots matched'),
            pass: true,
          }
        }

        const diffPath = join(
          diffDir,
          basename(name, { variant: r.state, idx: r.idx }),
        )

        if (r.diff) await writeFile(diffPath, r.diff)

        return {
          message: msgLine(
            r,
            `Screenshots did not match. If this was expected, set SPOTCHECK_UPDATE=true. Visual diff written to:\n\t${diffPath}`,
          ),
          pass: false,
        }
      }),
    )

    return {
      pass: matchResults.every(m => m.pass),
      message: () =>
        matchResults.reduce((msg, r) => {
          const fn = r.pass ? chalk.green : chalk.red
          const line = fn(r.message)

          return `${msg}\n${line}`
        }, `Screenshot comparison for "${name}" failed:`),
    }
  }
}
