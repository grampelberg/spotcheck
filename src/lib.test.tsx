import { afterAll, describe, expect, mock, test } from 'bun:test'
import { readdir, rm, stat } from 'node:fs/promises'
import { dirname, join } from 'node:path'

import { fileURLToPath } from 'bun'
import debug from 'debug'
import type { Pool } from 'generic-pool'
import type { Browser } from 'puppeteer'
import type { ReactElement } from 'react'

import { name } from '../package.json' with { type: 'json' }
import { env } from './env.ts'
import {
  __allStates,
  type Builder,
  type ScreenshotDiff,
  screenshot,
} from './lib.ts'
import { newPool } from './pool.ts'

// biome-ignore lint/correctness/noUnusedVariables: used in debug
const log: debug.Debugger = debug(`${name}:test:lib`)

const __dirname: string = dirname(fileURLToPath(import.meta.url))

const identicalResults = (results: ScreenshotDiff[]): void => {
  expect(results.every(r => r.identical)).toBe(true)
  expect(results.map(r => r.state)).toEqual(expect.arrayContaining(__allStates))
  expect(results.every(r => r.before)).toBeDefined()
  expect(results.every(r => r.diff)).toBeDefined()
}

const mtime = async (path: string): Promise<number> => {
  const { mtime } = await stat(path)
  return mtime.getTime()
}

describe('screenshot', async () => {
  const pool = newPool({ preserveBrowser: env().SPOTCHECK_PRESERVE })
  const css = ['* { margin: 0 }']
  const builder = async (): Promise<string[]> => css
  const opts = {
    update: env().SPOTCHECK_UPDATE,
  }
  const tmpPath = '__tmp__'
  const tmpDir = join(__dirname, tmpPath)

  if (env().TEST_CLEAN)
    afterAll(async () => await rm(tmpDir, { recursive: true, force: true }))

  test("doesn't call builder when there's no css", async () => {
    const mockBuilder = mock(builder)

    await screenshot(<div>Hello</div>, 'hello-nocss', mockBuilder, pool, {
      path: tmpPath,
    })

    expect(mockBuilder).not.toHaveBeenCalled()
  })

  describe.serial('works for JSX', async () => {
    const cssPaths = ['foobar', 'baz']
    const mockBuilder = mock(builder)

    const results = await screenshot(
      <div>Hello</div>,
      'hello',
      mockBuilder,
      pool,
      {
        css: cssPaths,
        ...opts,
      },
    )

    test('builder was called correctly', async () => {
      expect(mockBuilder).toHaveBeenCalledTimes(1)
      const call = mockBuilder.mock.calls[0] as [string, string[]] | undefined
      if (!call) throw new Error('Builder was not called')

      expect(call[0]).toEqual(expect.stringContaining('<html'))

      expect(call[0]).not.toContain('<style>')
      expect(call[1]).toEqual(expect.arrayContaining(cssPaths))

      expect(mockBuilder.mock.results[0]?.value).resolves.toEqual(
        expect.arrayContaining(css),
      )
    })

    test('results are correct', async () => {
      identicalResults(results)
    })
  })

  test('works for strings', async () => {
    const results = await screenshot(
      '<div>Hello</div>',
      'hello-str',
      builder,
      pool,
      { css: ['foobar'], ...opts },
    )

    identicalResults(results)
  })

  test.serial('throws when nothing is rendered', async () => {
    expect(async () =>
      screenshot('', 'hello-empty', builder, pool),
    ).toThrowErrorMatchingSnapshot()
  })

  test('injects css', async () => {
    const el = <div>Hello</div>
    const name = 'hello-css'
    const cssOpts = {
      states: ['default'],
      css: ['foobar'],
      path: '__tmp__',
    }
    await screenshot(el, name, async () => [], pool, cssOpts)

    const results = await screenshot(
      el,
      name,
      async () => ['* { color: white }'],
      pool,
      cssOpts,
    )

    expect(results.every(r => r.identical)).toBe(false)
  })

  describe.serial('file handling', async () => {
    const name = 'hello-path'
    const states = ['active', 'hover']
    const args: [ReactElement, string, Builder, Pool<Browser>] = [
      <>
        <div>Foo</div>
        <div>Bar</div>
      </>,
      name,
      builder,
      pool,
    ]
    const fileOpts = {
      states,
      path: tmpPath,
    }

    await screenshot(...args, fileOpts)

    const files = await readdir(tmpDir)
    const stateFiles = files.filter(n => n.includes(name))

    // Also tests writing files when none exist
    test('places screenshots where configured', async () => {
      expect(files.some(n => n.includes(name))).toBe(true)
    })

    test('only collects the states configured', async () => {
      expect(
        states.every(str => stateFiles.some(file => file.includes(str))),
      ).toBe(true)
    })

    test('takes a screenshot for each element', async () => {
      const elementIndex = ['0', '1']

      expect(
        elementIndex.every(str => stateFiles.some(file => file.includes(str))),
      ).toBe(true)
    })
  })

  test('updates screenshots when configured', async () => {
    const name = 'hello-update'
    const args: [ReactElement, string, Builder, Pool<Browser>] = [
      <div>Foo</div>,
      name,
      builder,
      pool,
    ]
    const fileOpts = {
      states: ['default'],
      path: tmpPath,
    }

    await screenshot(...args, fileOpts)

    const fname = (await readdir(tmpDir)).find(f => f.includes(name))

    expect(fname).toBeDefined()

    const fpath = join(tmpDir, fname as string)

    const startMtime = await mtime(fpath)

    await screenshot(...args, fileOpts)

    expect(startMtime).toBe(await mtime(fpath))

    await screenshot(...args, { ...fileOpts, update: true })
    expect(startMtime).toBeLessThan(await mtime(fpath))
  })

  describe('interaction reset', async () => {
    const name = 'hello-reset'
    const css = `
      * { background: blue }
      button:hover { background: red }
      button:active { border: 5px solid green }
      button:focus { color: white }
      `

    const args: [ReactElement, string, Builder, Pool<Browser>] = [
      <button type="button">Foo</button>,
      name,
      async (): Promise<string[]> => [css],
      pool,
    ]
    const fileOpts = {
      css: ['foobar'],
      path: tmpPath,
    }

    await screenshot(...args, fileOpts)

    test.each(__allStates)('%s is the same', async state => {
      const results = await screenshot(...args, {
        ...fileOpts,
        states: [state],
      })

      expect(results.every(r => r.identical)).toBe(true)
    })
  })

  test("doesn't wait for long animations", async () => {
    const name = 'hello-anim'
    const plain = `* { margin: 0; }`
    const animated = `
      * {
        margin: 0;
        opacity: 1;
        animation: fadeIn 1s ease-in forwards;
      }

      @keyframes fadeIn {
        from {
          opacity: 1;
        }
        to {
          opacity: 0;
        }
      }`
    const el = <div>foo</div>
    const animOpts = {
      css: ['foobar'],
      states: ['default'],
      path: tmpPath,
    }

    await screenshot(el, name, async () => [plain], pool, animOpts)

    const results = await screenshot(el, name, async () => [animated], pool, {
      ...animOpts,
    })

    expect(results.every(r => r.identical)).toBe(true)
  })

  // doesn't leak browser instances
  // has the correct visual elements
})
