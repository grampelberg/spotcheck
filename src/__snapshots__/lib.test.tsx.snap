// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`screenshot throws when nothing is rendered 1`] = `
"No elements found for "hello-empty":
<html lang="en">
  <head></head>
  <body>
    
  </html>
  "
`;
